cpp/src/lists/contains.cu

```diff
@@ -0,0 +1,279 @@
+          rmm::cuda_stream_view stream,
+          rmm::mr::device_memory_resource* mr) const
+  {
+    thrust::transform(rmm::exec_policy(stream),
+                      thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(d_lists.size()),
+                      output_bools.data<bool>(),
+                      [d_lists, search_key_iter] __device__(auto row_index) {
+                        auto search_key_and_validity    = search_key_iter[row_index];
+                        auto const& search_key_is_valid = search_key_and_validity.second;
+
+                        if (!search_key_is_valid) { return false; }
+
+                        auto list = cudf::list_device_view(d_lists, row_index);
+                        if (list.is_null()) { return false; }
+                        auto search_key = search_key_and_validity.first;
+
+                        return thrust::find_if(thrust::seq,
+                                               list.pair_begin<T>(),
+                                               list.pair_end<T>(),
+                                               [search_key] __device__(auto element_and_validity) {
+                                                 return element_and_validity.second &&
+                                                        (element_and_validity.first == search_key);
+                                               }) != list.pair_end<T>();
+                      });
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_numeric<T>() || cudf::is_chrono<T>() ||
+                     std::is_same<T, cudf::string_view>::value,
+                   void>
+  operator()(
+    cudf::detail::lists_column_device_view const& d_lists,
+    cudf::scalar const& search_key,
+    bool search_key_is_null,  // Ignored for scalars. This is decipherable from `search_key`.
+    cudf::mutable_column_device_view output_bools,
+    rmm::cuda_stream_view stream,
+    rmm::mr::device_memory_resource* mr) const
+  {
+    assert(search_key.is_valid() && "skey should have been checked for nulls by this point.");
+    return op_impl<T>(d_lists, detail::make_pair_iterator<T>(search_key), output_bools, stream, mr);
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_numeric<T>() || cudf::is_chrono<T>() ||
+                     std::is_same<T, cudf::string_view>::value,
+                   void>
+  operator()(cudf::detail::lists_column_device_view const& d_lists,
+             cudf::column_device_view const& d_search_keys,
+             bool search_keys_have_nulls,
+             cudf::mutable_column_device_view output_bools,
+             rmm::cuda_stream_view stream,
+             rmm::mr::device_memory_resource* mr) const
+  {
+    return search_keys_have_nulls
+             ? op_impl<T>(d_lists, d_search_keys.pair_begin<T, true>(), output_bools, stream, mr)
+             : op_impl<T>(d_lists, d_search_keys.pair_begin<T, false>(), output_bools, stream, mr);
+  }
```

Comment:
Lift the call to `pair_begin()` up to `contains_impl` so that you can collapse this functor down to a single `operator()`.

Reply:
> Lift the call to `pair_begin()` up to `contains_impl`.

Just a moment. In both cases, the pair iterators need the `<T>` to be available. Wouldn't that imply that we'd need to type_dispatch twice? First to fetch the iterators, and again to do the `contains()` work?

Reply:
Ugh, what a pain.

Reply:
>  It would all go a way if you could create a column_view of a scalar...

Meaning a `column_view` of `N` rows, backed by a scalar? (As opposed to a `column_view` of `1` row, which wouldn't be ideal.)
That would've been brilliant.

For the moment, I'll try hitting this by dispatching twice.

Reply:
Sorry, I tried this. It doesn't seem to reduce the code much. (We land up transplanting the code where the pair iterators are constructed to another location.) This doesn't look like an improvement. This is disappointing. :/

Permission to punt this until we write an abstraction for non-typed iterators that are uniform between `scalar` and `column_view`?

Reply:
(I realize that @harrism has already :+1:-ed this. I thought I'd exhaust our options before checking this in.)